File: kotlin/src/main/kotlin/adventofcode2022/Day09.kt
package adventofcode2022

import adventofcode2022.util.runWithStopwatch
import java.io.File
import kotlin.math.abs

fun main() {
    runWithStopwatch { part1() }
    runWithStopwatch { part2() }
}

private fun part1() {
    val commands = File("./input/day09.txt").readLines()
    val rope = Rope(1)
    val tailLocationHistory = execute(commands, rope)
    println("Part1: ${tailLocationHistory.size}")
}

private fun part2() {
    val commands = File("./input/day09.txt").readLines()
    val rope = Rope(9)
    val tailLocationHistory = execute(commands, rope)
    println("Part2: ${tailLocationHistory.size}")
}

private fun execute(commands: List<String>, rope: Rope, ): MutableSet<Pair<Int, Int>> {
    val tailLocationHistory = mutableSetOf<Pair<Int, Int>>()
    for (commandText in commands) {
        val command = commandText.split(" ")
        for (ignore in 0 until command[1].toInt()) {
            when (command[0]) {
                "U" -> rope.moveUp()
                "D" -> rope.moveDown()
                "L" -> rope.moveLeft()
                "R" -> rope.moveRight()
                else -> error("Strange input!")
            }
            tailLocationHistory.add(rope.last())
        }
    }
    return tailLocationHistory
}

class Rope(size: Int) {
    private var head: Pair<Int, Int> = 0 to 0
    private var tail: Array<Pair<Int, Int>>

    init {
        tail = Array(size) { 0 to 0 }
    }

    fun moveRight() {
        val x = head.first + 1
        val y = head.second
        head = x to y
        updateTail()
    }

    fun moveLeft() {
        val x = head.first - 1
        val y = head.second
        head = x to y
        updateTail()
    }

    fun moveUp() {
        val x = head.first
        val y = head.second + 1
        head = x to y
        updateTail()
    }

    fun moveDown() {
        val x = head.first
        val y = head.second - 1
        head = x to y
        updateTail()
    }

    fun last(): Pair<Int, Int> {
        return tail.last()
    }

    private fun updateTail() {
        var tempHead = head
        for (i in tail.indices) {
            val (dx, dy) = getDiff(tempHead, tail[i])
            if (!isAdjacent(dx, dy)) {
                tail[i] = tail[i].first + sign(dx) to tail[i].second + sign(dy)
            }
            tempHead = tail[i]
        }
    }

    private fun getDiff(first: Pair<Int, Int>, second: Pair<Int, Int>): Pair<Int, Int> {
        val dx = first.first - second.first
        val dy = first.second - second.second
        return dx to dy
    }

    private fun sign(value: Int): Int {
        return if (value > 0) {
            1
        } else if (value < 0) {
            -1
        } else {
            0
        }
    }

    private fun isAdjacent(dx: Int, dy: Int): Boolean {
        return abs(dx) <= 1 && abs(dy) <= 1
    }
}

File: python/Day09.py
from utils.Stopwatch import runWithStopwatch


def adjacent(dx, dy):
    return abs(dx) <= 1 and abs(dy) <= 1


def sign(val):
    if val > 0:
        return 1
    elif val < 0:
        return -1
    else:
        return 0


def update_tail(new_head, tail):
    dx = new_head[0] - tail[0]
    dy = new_head[1] - tail[1]
    if not adjacent(dx, dy):
        tail[0] += sign(dx)
        tail[1] += sign(dy)
    return tail


def moveUp(head, tail):
    new_head = [head[0], head[1] + 1]
    tail = update_tail(new_head, tail)
    return new_head, tail


def moveDown(head, tail):
    new_head = [head[0], head[1] - 1]
    tail = update_tail(new_head, tail)
    return new_head, tail


def moveLeft(head, tail):
    new_head = [head[0] - 1, head[1]]
    tail = update_tail(new_head, tail)
    return new_head, tail


def moveRight(head, tail):
    new_head = [head[0] + 1, head[1]]
    tail = update_tail(new_head, tail)
    return new_head, tail


def part1():
    commands = open("../input/day09.txt", "r").readlines()
    tail_position_history = []
    head = [0, 0]
    tail = [0, 0]
    for command in commands:
        command = command.strip().split(" ")
        for i in range(0, int(command[1])):
            match command[0]:
                case "U":
                    head, tail = moveUp(head, tail)
                case "D":
                    head, tail = moveDown(head, tail)
                case "L":
                    head, tail = moveLeft(head, tail)
                case "R":
                    head, tail = moveRight(head, tail)
            tail_position_history.append(tuple(tail))

    print("Number of positions visited at least once: " + str(len(set(tail_position_history))))


def part2():


if __name__ == '__main__':
    runWithStopwatch(part1)
    runWithStopwatch(part2)
